README.md
# REANA example - CMS Reconstruction

[![image](https://img.shields.io/pypi/pyversions/reana-demo-cms-reco.svg)](https://pypi.org/pypi/reana-demo-cms-reco)
[![image](https://img.shields.io/badge/discourse-forum-blue.svg)](https://forum.reana.io)
[![image](https://img.shields.io/github/license/reanahub/reana.svg)](https://github.com/reanahub/reana-demo-cms-reco/blob/master/LICENSE)

## About

This REANA reproducible analysis example demonstrates the reconstruction
procedure of the CMS collaboration from [raw
data](http://opendata.cern.ch/search?page=1&size=20&experiment=CMS&file_type=raw)
to [Analysis Object Data
(AOD)](https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookDataFormats#AoD),
for the year `2011` and the data set `DoubleElectron`.

The workflow consists of the steps need for the samples reconstruction, as taken
from the [CMS legacy validation
repo](https://github.com/cms-legacydata-validation/RAWToAODValidation/tree/master).

## Reconstruction procedure

### 1 & 2. Input data and Analysis code

Any raw input data from the
[CERN open data
platform](http://opendata.cern.ch/search?page=1&size=20&experiment=CMS&type=Dataset&subtype=Collision&subtype=Derived&subtype=Simulated&file_type=raw)
should be valid for reconstruction. In this example, the input is taken from:
`root://eospublic.cern.ch//eos/opendata/cms/Run2011A/DoubleElectron/RAW/v1/000/160/433/C046161E-0D4E-E011-BCBA-0030487CD906.root`

The reconstruction step can be repeated with a configuration file that depends
on the analyzed data, e.g. [this example](http://opendata.cern.ch/record/43), or
by creating our own configuration file (created in a CMS VM) and then changing
the script accordingly:

```console
cmsDriver.py reco -s RAW2DIGI,L1Reco,RECO,USER:EventFilter/HcalRawToDigi/hcallaserhbhehffilter2012_cff.hcallLaser2012Filter --data --conditions FT_R_53_LV5::All --eventcontent AOD --customise Configuration/DataProcessing/RecoTLR.customisePrompt --no_exec --python reco_cmsdriver2011.py
```

### 3. Compute environment

In order to be able to rerun the analysis even several years in the future, we
need to "encapsulate the current compute environment", for example to freeze the
software package versions our analysis is using. We shall achieve this by
preparing a [Docker](https://www.docker.com/) container image for our analysis
steps.

This analysis example runs within the [CMSSW](http://cms-sw.github.io/) analysis
framework that was packaged for Docker in
[cmsopendata](https://hub.docker.com/u/cmsopendata). The different images
corresponds to data sets taken in different years. Instructions can be found
under [this repo](http://opendata.cern.ch/docs/cms-guide-docker).

Moreover, the re-reconstruction task needs access run-time to the condition
database and inside a [CMS
VM](http://opendata.cern.ch/search?page=1&size=20&q=virtual%20machine&subtype=VM&type=Environment&experiment=CMS),
this is achieved with the commands:

```console
$ ln -sf /cvmfs/cms-opendata-conddb.cern.ch/FT_53_LV5_AN1_RUNA FT_53_LV5_AN1
$ ln -sf /cvmfs/cms-opendata-conddb.cern.ch/FT_53_LV5_AN1_RUNA.db FT_53_LV5_AN1_RUNA.db
```

For REANA, the condition database on CVMFS can be accessed with any container,
the only requirement is that the user should specify the necessary CVMFS volumes
to be live-mounted in the `reana.yaml` resource section, as described
[on the
web](https://reana.readthedocs.io/en/latest/userguide.html#declare-necessary-resources).

### 4. Analysis workflow

First, we have to set up the environment variables accordingly for the
[CMS SW](http://cms-sw.github.io/). Although this is done in the docker image,
REANA overrides them and they need to be reset. This is done by copying the [cms
entrypoint.sh
script](https://github.com/clelange/cmssw-docker/blob/master/standalone/entrypoint.sh):

```console
$ source /opt/cms/cmsset_default.sh
$ scramv1 project CMSSW CMSSW_5_3_32
$ cd CMSSW_5_3_32/src
$ eval `scramv1 runtime -sh`
```

The actual commands that are needed to carry out the analysis in the CMS
specific environment are then:

```console
$ ln -sf /cvmfs/cms-opendata-conddb.cern.ch/FT_53_LV5_AN1_RUNA FT_53_LV5_AN1
$ ln -sf /cvmfs/cms-opendata-conddb.cern.ch/FT_53_LV5_AN1_RUNA.db FT_53_LV5_AN1_RUNA.db
$ ls -l
$ ls -l /cvmfs/
$ cmsRun reco_cmsdriver2011.py
```

This demo represents a "workflow factory" script that will produce REANA
workflows for given parameters for the CMS RAW to AOD reconstruction procedure.

Following successful tests (see other branches), we know that REANA is able to
run CMS reconstruction for a variety of RAW samples (e.g. dataset SingleMu) and
data-taking years (e.g. 2011).

### Example

Before running example, you might want to install necessary packages:

```console
$ # create new virtual environment
$ virtualenv ~/.virtualenvs/myreana
$ source ~/.virtualenvs/myreana/bin/activate
$ # install reana-commons and reana-client
$ pip install git+git://github.com/reanahub/reana-demo-cms-reco.git@master#egg=cms-reco
```

After, the following will generate the workflow to run the example for a given
record id, with its metadata retrieved using the [COD
Client](https://github.com/cernopendata/cernopendata-client). This generates a
workflow in a given output directory, where the `reana.yaml` file lives with all
necessary inputs:

```console
$ cernopendata-client get-record --recid 39 | tee cms-reco-config.json
# # use the values from the 'cms-reco-config.json' file
$ cms-reco --create-workflow
    Created `cms-reco-SingleElectron-2011` directory.
$ cd cms-reco-SingleElectron-2011
$ reana-client run
```


CHANGES.rst
Changes
=======

Version 0.1.0 (2019-MM-DD)
--------------------------
.. TODO add realease date

- Initial public release.

.. admonition:: Please beware

   Please note that cms-reco is in an early alpha stage of
   its development. The developer preview releases are meant for early
   adopters and testers. Please don't rely on released versions for any
   production purposes yet.

run-tests.sh
#!/bin/sh
#
# This file is part of REANA.
# Copyright (C) 2019 CERN.
#
# REANA is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

#ToDo: use the branches for different years as tests cases

pydocstyle cms_reco && \
isort -rc -c -df **/*.py && \
cms-reco --help > cmd_list.txt && \
diff -q -w docs/cmd_list.txt cmd_list.txt  && \
rm cmd_list.txt && \
sphinx-build -qnNW docs docs/_build/html && \
python setup.py test && \
sphinx-build -qnNW -b doctest docs docs/_build/doctest

tests/test_workflow.py
# -*- coding: utf-8 -*-
#
# This file is part of reana.
# Copyright (C) 2019 CERN.
#
# REANA is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.


import filecmp
import os
import subprocess as sp
import urllib.request as ur

from cms_reco.utils import remove_additionally_generated_files, remove_folder


def generate_file(recid=46, directory='test'):
    """Generate the reana.yaml using the workflow factory."""
    _cms_reco_client = "cms-reco"

    sp.call("{0} load-config --recid {1} "
            "&& {0} create-workflow --directory {2} "
            .format(_cms_reco_client, recid, directory),
            shell=True)
    workflow_file = "{0}/reana.yaml".format(directory)
    if os.path.isfile(workflow_file):
        return workflow_file
    else:
        return None


def download_existing_file(url, local_file_name='reference.yaml'):
    """Get a well known and working reana.yaml workflow file."""
    ur.urlretrieve(url, local_file_name)
    return local_file_name


def compare_files(ref_file, gen_file):
    """Compare two files for their content."""
    return filecmp.cmp(ref_file, "{0}".format(gen_file), shallow=False)


def test_workflow_2011():
    """Test for the year 2011."""
    ref_file = download_existing_file(
        'https://raw.githubusercontent.com/dprelipcean/'
        'reana-demo-cms-reco/reana_example_2011/reana.yaml')

    tmp_folder = "test"
    gen_file = generate_file(recid=46, directory=tmp_folder)

    assert compare_files(ref_file, gen_file)

    remove_additionally_generated_files(ref_file)
    remove_folder(tmp_folder)


cms_reco/cli.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# This file is part of reana.
# Copyright (C) 2019 CERN.
#
# REANA is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

"""Reana workflow factory cli."""


import os

import click
from cookiecutter.exceptions import OutputDirExistsException
from cookiecutter.main import cookiecutter

from .utils import (get_config_from_json, load_config_from_cod, run_analysis,
                    run_pipeline, valid_run_years, workflow_engines)


@click.group()
def cms_reco():
    """Info."""
    pass


@click.option('--config_file',
              default='cms_reco/cms-reco-config.json',
              help='recid for the data set to be reconstructed')
@click.option('--recid',
              help='recid for the data set to be reconstructed')
@cms_reco.command()
def load_config(recid, config_file):
    """Download config file using the cern open data client."""
    load_config_from_cod(recid, config_file)
    print("Downloaded config file from cod as {}.".format(config_file))


@click.option('--directory',
              default='',
              help='directory for the analysis to be executed')
@click.option('--workflow_engine',
              default='serial',
              help='workflow engine to be used',
              type=click.Choice(workflow_engines))
@click.option('--nevents',
              default='1',  # ToDo: change to "-1" for full data set reco
              help='number of events to be reconstructed')
@click.option('--dataset',
              default="DoubleElectron",
              help='data set to be reconstructed')
@click.option('--year',
              default="2011",
              help='year the data set was recorded',
              type=click.Choice(valid_run_years)
              )
@cms_reco.command()
def create_workflow(year, dataset, nevents, workflow_engine, directory):
    """Create workflow from given parameters."""
    template_path = "{}/cms_reco/cookiecutter_template/workflow_factory/{}"\
        .format(os.getcwd(), workflow_engine)

    config = get_config_from_json()

    # Set additional configs
    if nevents:
        config['nevents'] = nevents
    if directory:
        config['directory_name'] = directory

    try:
        cookiecutter(template_path, no_input=True, extra_context=config)
    except OutputDirExistsException:
        print("Output Directory already exists, please choose a different name \
        or rename the existing one.")
    print("Created `{0}` directory.".format(config["directory_name"]))


cms_reco/utils.py
# -*- coding: utf-8 -*-
#
# This file is part of reana.
# Copyright (C) 2019 CERN.
#
# REANA is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

"""Reana workflow factory helper functions."""

import json
import os
import sys
import shutil
import subprocess as sp
import urllib.request as ur

import jq

with open("{0}/cms_reco/cod-validation.json".format(os.getcwd())) \
        as validation_file:
    validation_data = json.load(validation_file)

    valid_run_years = validation_data["years"]
    workflow_engines = validation_data["workflow_engines"]


def get_config_from_json():
    """Get the needed configuration variables from the COD client config."""
    with open('{0}/cms_reco/cms-reco-config.json'.format(os.getcwd())) \
            as config_file:
        data = json.load(config_file)

        return {'directory_name': custom_directory_name(data),
                'year': get_year(data),
                'cmssw_version': get_cms_release(data),
                'global_tag': get_global_tag(data),
                'dataset_file': get_dataset(data)
                }


def get_global_tag(data):
    """Get the global tag for the CMS cond db."""
    return jq.jq(".metadata.system_details.global_tag").transform(data)


def get_cms_release(data):
    """Get the CMS SW release version."""
    # keep only the version number
    return jq.jq(".metadata.system_details.release").transform(data)[6:]


def download_index_file(data, local_file_name, file_format="txt"):
    """Download the index file from COD platform."""
    recid = get_recid(data)
    url = get_index_file_name(data, recid, file_format)
    try:
        if not os.path.isfile(local_file_name):
            ur.urlretrieve(url, local_file_name)

    except Exception as e:
        print("failed to download index file: {0}".format(e))


def remove_additionally_generated_files(files):
    """Remove files that the end user does not need."""
    if type(files) in (tuple, list):
        for file in files:
            remove_additionally_generated_files(file)
    elif type(files) == str:
        if os.path.isfile(files):
            os.remove(files)
        else:
            raise Exception("Error: {} file not found".format(files))
    else:
        raise Exception("Files type {} not recognized.".format(type(files)))


def remove_folder(mydir):
    """Remove folder."""
    try:
        shutil.rmtree(mydir)
    except OSError as e:
        print("Error: {} - {}.".format(e.filename, e.strerror))


def get_index_file_name(data, recid, file_format):
    """Get the dataset specific index file name."""
    if file_format == "txt":
        index_file = jq.jq(".metadata._files").transform(data)[1]["key"]
        url = "http://opendata.cern.ch/record/{0}//files/{1}" \
            .format(recid, index_file)
        return url


def get_dataset(data, local_file_name="./index.txt"):
    """Get a data set file name from the index file."""
    try:
        download_index_file(data, local_file_name)
        dataset = choose_dataset_from_file(local_file_name)
    finally:
        remove_additionally_generated_files(local_file_name)

    return dataset


def choose_dataset_from_file(local_file_name):
    """Chose a specific dataset from file."""
    with open("{0}/{1}".format(os.getcwd(), local_file_name)) as file:
        # use the first dataset in the index file
        dataset = file.readline()
    return dataset


def get_recid(data):
    """Get the record id."""
    return jq.jq(".id").transform(data)


def get_title(data):
    """Get the data set title."""
    return jq.jq(".metadata.title").transform(data)


def get_year(data):
    """Get creation year for the data set."""
    return jq.jq(".metadata.date_created").transform(data)[0]


def get_name_from_title(title):
    """Get the data set name from the title."""
    return os.path.dirname(os.path.dirname(title))[1:]


def custom_directory_name(data):
    """Return a custom directory name based on the title."""
    return "cms-reco-{}-{}".format(get_name_from_title(get_title(data)),
                                   get_year(data))


def load_config_from_cod(recid, config_file):
    """Get the config file using cern open data client."""
    _cod_client = "cernopendata-client"
    _get_config_cmd = "get-record --recid"
    sp.call("{0} {1} {2} | tee {3}"
            .format(_cod_client, _get_config_cmd, recid, config_file),
            shell=True)


def run_analysis(directory_name):
    """Run the analysis on reana."""
    sp.call("cd {0} && reana-client run reana.yaml"
            .format(directory_name),
            shell=True)


def run_pipeline(recid, directory):
    """Run the full pipeline."""
    _cms_reco_client = "cms-reco"
    sp.call("{0} load-config --recid {1} "
            "&& {0} create-workflow --directory {2} "
            "&& {0} run-reco --directory {2}"
            .format(_cms_reco_client, recid, directory),
            shell=True)


cms_reco/cookiecutter_template/workflow_factory/cwl/{{cookiecutter.directory_name}}/reco_cmsdriver{{cookiecutter.year}}.py
# Auto generated configuration file
# using:
# Revision: 1.381.2.28
# Source: /local/reps/CMSSW/CMSSW/Configuration/PyReleaseValidation/python/ConfigBuilder.py,v
# with command line options: reco -s RAW2DIGI,L1Reco,RECO,USER:EventFilter/HcalRawToDigi/hcallaserhbhehffilter2012_cff.hcallLaser2012Filter --data --conditions FT_R_53_LV5::All --eventcontent AOD --customise Configuration/DataProcessing/RecoTLR.customisePrompt --no_exec --python reco_cmsdriver2011.py
import FWCore.ParameterSet.Config as cms

process = cms.Process('RECO')

# import of standard configurations
process.load('Configuration.StandardSequences.Services_cff')
process.load('SimGeneral.HepPDTESSource.pythiapdt_cfi')
process.load('FWCore.MessageService.MessageLogger_cfi')
process.load('Configuration.EventContent.EventContent_cff')
process.load('Configuration.StandardSequences.GeometryRecoDB_cff')
process.load('Configuration.StandardSequences.MagneticField_AutoFromDBCurrent_cff')
process.load('Configuration.StandardSequences.RawToDigi_Data_cff')
process.load('Configuration.StandardSequences.L1Reco_cff')
process.load('Configuration.StandardSequences.Reconstruction_Data_cff')
process.load('EventFilter.HcalRawToDigi.hcallaserhbhehffilter2012_cff')
process.load('Configuration.StandardSequences.EndOfProcess_cff')
process.load('Configuration.StandardSequences.FrontierConditions_GlobalTag_cff')

process.maxEvents = cms.untracked.PSet(
    input = cms.untracked.int32({{cookiecutter.nevents}})
)

# Input source
process.source = cms.Source("PoolSource",
    secondaryFileNames = cms.untracked.vstring(),
    fileNames = cms.untracked.vstring(
        {{cookiecutter.dataset_file}}
                                      )
)

process.options = cms.untracked.PSet(

)

# Production Info
process.configurationMetadata = cms.untracked.PSet(
    version = cms.untracked.string('$Revision: 1.381.2.28 $'),
    annotation = cms.untracked.string('reco nevts:1'),
    name = cms.untracked.string('PyReleaseValidation')
)

# Output definition

process.AODoutput = cms.OutputModule("PoolOutputModule",
    eventAutoFlushCompressedSize = cms.untracked.int32(15728640),
    outputCommands = process.AODEventContent.outputCommands,
    fileName = cms.untracked.string('reco_RAW2DIGI_L1Reco_RECO_USER.root'),
    dataset = cms.untracked.PSet(
        filterName = cms.untracked.string(''),
        dataTier = cms.untracked.string('')
    )
)

# Additional output definition

#Additional statements
process.GlobalTag.connect = cms.string('sqlite_file:/cvmfs/cms-opendata-conddb.cern.ch/{{cookiecutter.global_tag}}_RUNA.db')
process.GlobalTag.globaltag = '{{cookiecutter.global_tag}}::All'

# Path and EndPath definitions
process.raw2digi_step = cms.Path(process.RawToDigi)
process.L1Reco_step = cms.Path(process.L1Reco)
process.reconstruction_step = cms.Path(process.reconstruction)
process.user_step = cms.Path(process.hcallLaser2012Filter)
process.endjob_step = cms.EndPath(process.endOfProcess)
process.AODoutput_step = cms.EndPath(process.AODoutput)

# Schedule definition
process.schedule = cms.Schedule(process.raw2digi_step,process.L1Reco_step,process.reconstruction_step,process.user_step,process.endjob_step,process.AODoutput_step)

# customisation of the process.

# Automatic addition of the customisation function from Configuration.DataProcessing.RecoTLR
from Configuration.DataProcessing.RecoTLR import customisePrompt

#call to customisation function customisePrompt imported from Configuration.DataProcessing.RecoTLR
process = customisePrompt(process)

# End of customisation functions

reco_cmsdriver2011.py
# Auto generated configuration file
# using:
# Revision: 1.381.2.28
# Source: /local/reps/CMSSW/CMSSW/Configuration/PyReleaseValidation/python/ConfigBuilder.py,v
# with command line options: reco -s RAW2DIGI,L1Reco,RECO,USER:EventFilter/HcalRawToDigi/hcallaserhbhehffilter2012_cff.hcallLaser2012Filter --data --conditions FT_R_53_LV5::All --eventcontent AOD --customise Configuration/DataProcessing/RecoTLR.customisePrompt --no_exec --python reco_cmsdriver2011.py
import FWCore.ParameterSet.Config as cms

process = cms.Process('RECO')

# import of standard configurations
process.load('Configuration.StandardSequences.Services_cff')
process.load('SimGeneral.HepPDTESSource.pythiapdt_cfi')
process.load('FWCore.MessageService.MessageLogger_cfi')
process.load('Configuration.EventContent.EventContent_cff')
process.load('Configuration.StandardSequences.GeometryRecoDB_cff')
process.load('Configuration.StandardSequences.MagneticField_AutoFromDBCurrent_cff')
process.load('Configuration.StandardSequences.RawToDigi_Data_cff')
process.load('Configuration.StandardSequences.L1Reco_cff')
process.load('Configuration.StandardSequences.Reconstruction_Data_cff')
process.load('EventFilter.HcalRawToDigi.hcallaserhbhehffilter2012_cff')
process.load('Configuration.StandardSequences.EndOfProcess_cff')
process.load('Configuration.StandardSequences.FrontierConditions_GlobalTag_cff')

process.maxEvents = cms.untracked.PSet(
    input = cms.untracked.int32(-1)
)

# Input source
process.source = cms.Source("PoolSource",
    secondaryFileNames = cms.untracked.vstring(),
    fileNames = cms.untracked.vstring(
        'root://eospublic.cern.ch//eos/opendata/cms/Run2011A/DoubleElectron/RAW/v1/000/160/433/C046161E-0D4E-E011-BCBA-0030487CD906.root'
                                      )
)

process.options = cms.untracked.PSet(

)

# Production Info
process.configurationMetadata = cms.untracked.PSet(
    version = cms.untracked.string('$Revision: 1.381.2.28 $'),
    annotation = cms.untracked.string('reco nevts:1'),
    name = cms.untracked.string('PyReleaseValidation')
)

# Output definition

process.AODoutput = cms.OutputModule("PoolOutputModule",
    eventAutoFlushCompressedSize = cms.untracked.int32(15728640),
    outputCommands = process.AODEventContent.outputCommands,
    fileName = cms.untracked.string('reco_RAW2DIGI_L1Reco_RECO_USER.root'),
    dataset = cms.untracked.PSet(
        filterName = cms.untracked.string(''),
        dataTier = cms.untracked.string('')
    )
)

# Additional output definition

#Additional statements
process.GlobalTag.connect = cms.string('sqlite_file:/cvmfs/cms-opendata-conddb.cern.ch/FT_53_LV5_AN1_RUNA.db')
process.GlobalTag.globaltag = 'FT_53_LV5_AN1::All'

# Path and EndPath definitions
process.raw2digi_step = cms.Path(process.RawToDigi)
process.L1Reco_step = cms.Path(process.L1Reco)
process.reconstruction_step = cms.Path(process.reconstruction)
process.user_step = cms.Path(process.hcallLaser2012Filter)
process.endjob_step = cms.EndPath(process.endOfProcess)
process.AODoutput_step = cms.EndPath(process.AODoutput)

# Schedule definition
process.schedule = cms.Schedule(process.raw2digi_step,process.L1Reco_step,process.reconstruction_step,process.user_step,process.endjob_step,process.AODoutput_step)

# customisation of the process.

# Automatic addition of the customisation function from Configuration.DataProcessing.RecoTLR
from Configuration.DataProcessing.RecoTLR import customisePrompt

#call to customisation function customisePrompt imported from Configuration.DataProcessing.RecoTLR
process = customisePrompt(process)

# End of customisation functions
